pkg/templates/nstemplatetiers/nstemplatetier_generator.go

```diff
@@ -0,0 +1,209 @@
+  name: {{ .Name }}
+spec:
+  namespaces:{{ range $index, $ns := .Namespaces }}
+  - type: {{ .Type }}
+    revision: {{ .Revision }}
+    template:
+{{ .Template }}{{ end }}`)
+	if initTmplErr != nil {
+		log.Error(initTmplErr, "failed to initialize NSTemplateTier template")
+	}
+}
+
+// NSTemplateTierGenerator the NSTemplateTier manifest generator
+type NSTemplateTierGenerator struct {
+	asset     func(name string) ([]byte, error) // the func which gives access to the
+	revisions map[string]map[string]string
+}
+
+// NewNSTemplateTierGenerator returns a new NSTemplateTierGenerator
+func NewNSTemplateTierGenerator(asset func(name string) ([]byte, error)) (*NSTemplateTierGenerator, error) {
+	metadata, err := asset("metadata.yaml")
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to initialize the NSTemplateTierGenerator")
+	}
+	revisions, err := parseAllRevisions(metadata)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to initialize the NSTemplateTierGenerator")
+	}
+	return &NSTemplateTierGenerator{
+		asset:     asset,
+		revisions: revisions,
+	}, nil
+}
+
+// parseRevisions returns a "supermap" in which:
+// - each key is a tier kind (eg: "basic", "advanced", etc.)
+// - each value is a map of revisions indexed by their associated template kind
+//   (eg: {"code":"123456", "dev":"abcdef", "stage":"cafe01"})
+func parseAllRevisions(metadata []byte) (map[string]map[string]string, error) {
+	data := make(map[string]interface{})
+	err := yaml.Unmarshal([]byte(metadata), &data)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse all template revisions")
+	}
+	// reads all entries in the YAML, splitting the keys
+	// as we expect the following format: '<tier_kind>-<namespace_kind>' (eg: 'advanced-code')
+	revisions := make(map[string]map[string]string)
+	for filename, revision := range data {
+		data := strings.Split(filename, "-")
+		if len(data) != 2 {
+			log.Info("invalid namespace template filename. Expected format: '<tier_kind>-<namespace_kind>'", "filename", filename)
+			continue
+		}
+		tierKind := data[0]
+		nsKind := data[1]
+		// create a new entry if needed
+		revision, ok := revision.(string)
+		if !ok {
+			log.Info("invalid namespace template filename revision. Expected a string", "filename", filename, "revision", revision)
+			continue
+		}
+		if _, ok := revisions[tierKind]; !ok {
+			revisions[tierKind] = make(map[string]string, 3) // expect 3 entries: 'code', 'dev' and 'stage'
+		}
+		revisions[tierKind][nsKind] = revision
+	}
+	log.Info("templates revisions loaded", "revisions", revisions)
+	return revisions, nil
+}
+
+// GenerateAllManifests generates all manifests, indexed by their associated tier kind and by their namespace kind
+// eg:
+// - advanced:
+//   - code: <[]byte>
+//   - dev: <[]byte>
+//   - stage: <[]byte>
+// - basic:
+//   - code: <[]byte>
+//   - dev: <[]byte>
+//   - stage: <[]byte>
+func (g NSTemplateTierGenerator) GenerateAllManifests() (map[string][]byte, error) {
```

Comment:
thanks for your feedback. Indeed, it's going to be more straightforward if we directly build the objects instead of generating a large YAML file here, and parse it in a subsequent PR. So let me refactor this PR to adapt the code to your suggestions.